x-pack/plugin/repository-encrypted/src/test/java/org/elasticsearch/repositories/encrypted/ChainingInputStreamTests.java

```diff
@@ -0,0 +1,573 @@
+            test3.skip(Randomness.get().nextInt(32));
+        });
+        ChainingInputStream test4 = newEmptyStream();
+        test4.close();
+        e = expectThrows(IOException.class, () -> {
+            test4.available();
+        });
+        ChainingInputStream test5 = newEmptyStream();
+        test5.close();
+        e = expectThrows(IOException.class, () -> {
+            test5.reset();
+        });
+        ChainingInputStream test6 = newEmptyStream();
+        test6.close();
+        test6.mark(Randomness.get().nextInt());
+    }
+
+    public void testHeadComponentArgumentIsNull() throws Exception {
+        AtomicReference<InputStream> headInputStream = new AtomicReference<>();
+        AtomicBoolean nextCalled = new AtomicBoolean(false);
+        ChainingInputStream test = new ChainingInputStream() {
+            @Override
+            InputStream nextComponent(InputStream currentComponentIn) throws IOException {
+                headInputStream.set(currentComponentIn);
+                nextCalled.set(true);
+                return null;
+            }
+        };
+        assertThat(test.read(), Matchers.is(-1));
+        assertThat(nextCalled.get(), Matchers.is(true));
+        assertThat(headInputStream.get(), Matchers.nullValue());
+    }
+
+    public void testChaining() throws Exception {
+        int componentCount = 2 + Randomness.get().nextInt(8);
+        TestInputStream[] sourceComponents = new TestInputStream[componentCount];
+        for (int i = 0; i < sourceComponents.length; i++) {
+            byte[] b = new byte[Randomness.get().nextInt(2)];
+            Randomness.get().nextBytes(b);
+            sourceComponents[i] = new TestInputStream(b);
+        }
+        ChainingInputStream test = new ChainingInputStream() {
+            int i = 0;
+            @Override
+            InputStream nextComponent(InputStream currentComponentIn) throws IOException {
+                if (i == 0) {
+                    assertThat(currentComponentIn, Matchers.nullValue());
+                    return sourceComponents[i++];
+                } else if (i < sourceComponents.length) {
+                    assertThat(((TestInputStream) currentComponentIn).closed.get(), Matchers.is(true));
+                    assertThat(currentComponentIn, Matchers.is(sourceComponents[i-1]));
+                    return sourceComponents[i++];
+                } else if (i == sourceComponents.length) {
+                    assertThat(((TestInputStream) currentComponentIn).closed.get(), Matchers.is(true));
+                    assertThat(currentComponentIn, Matchers.is(sourceComponents[i-1]));
+                    i++;
+                    return null;
+                } else {
+                    throw new IllegalStateException();
+                }
+            }
+
+            @Override
+            public boolean markSupported() {
+                return false;
+            }
+        };
+        test.readAllBytes();
```

Comment:
Shouldn't we check that this returns the correct sequence of bytes?
You could have a single `ByteBuffer`, and call `buffer.put(b)` in the loop that builds the test streams.
Then here, `assertThat(test.readAllBytes(), equalTo(buffer.array()));`

Reply:
Thanks for the suggestion.